### ISSUE TITLE
store both mmr index and block height into database for output

### ISSUE THREAD

**Issue by garyyu** [mimblewimble/grin#2889]
Currently, both `/v1/txhashset/outputs` API and `/v1/chain/outputs/byids` API are heavily used in wallet, and both API are very slow.

The heaviest part of these 2 API are `get_header_for_output` which gets the block header from an output commit.

https://github.com/mimblewimble/grin/blob/a89a1024b6dd7cc5d3def6b41edae0e46b55658d/chain/src/chain.rs#L1225-L1257

which has to do a loop searching to find the right height, and each loop contains two `get_header_by_height` function call.

The loop count depends on which commit we're querying, but it's normal case to get Log2(height) loops.  For example:  if the final result height is `159538`, then we need `16` loops, and that means `32` calls to `get_header_by_height`.

And  that `get_header_by_height` is not a light-cost function call:
https://github.com/mimblewimble/grin/blob/a89a1024b6dd7cc5d3def6b41edae0e46b55658d/chain/src/txhashset/txhashset.rs#L229-L236

---

But we have a simple way to change that and get a much faster improvement on these two APIs.

The way is to have both mmr index and block height saving to database for output.

Currently, when apply a new block, we save the output position (MMR index) into the database:
```rust
	pub fn apply_block(&mut self, b: &Block) -> Result<(), Error> {
		self.apply_header(&b.header)?;

		for out in b.outputs() {
			let pos = self.apply_output(out)?;
			// Update the output_pos index for the new output.
			self.batch.save_output_pos(&out.commitment(), pos)?;   <<<< Here !
		}
		...
```
If we change this `pos` to `(pos, b.header.height)`, then all those pains will be gone :-)

(but we have a pain to upgrade the database for the compatibility in one new version which take this improvement)

And Thanks @DavidBurkett 's discussion on https://github.com/GrinPlusPlus/GrinPlusPlus/issues/63#issuecomment-500658757 which leads me to realize this performance problem.

**Comment by antiochp** [comment]
:+1: Makes a lot of sense to me to store these in the db.

**Comment by antiochp** [comment]
For backward compatibility and ease of migration we might want to consider adding an additional index and avoid needing to migrate the existing one - 

```
A) commitment -> output_pos
B) output_pos -> block_height
```

Index (A) has the benefit of minimal chance of key collision due to using output commitments.
As long as index (B) was used in conjunction with (A) and not in isolation then we should be fine.
We could not necessarily rely on the `height` for a given `pos` without indexing via the `commitment` first (in the case of a fork/reorg etc.)

The overhead of two associated indexes would be relatively small compared to a single index (2 lookups vs single lookup and 8 bytes of additional data per entry).

----

That said - on node startup it should be relatively straightforward to read the index and migrate it to the new data format if necessary.

**Comment by garyyu** [comment]
> Index (A) has the benefit of minimal chance of key collision due to output commitments as keys.

Could you remind me why we need this "minimal chance of key collision"?

**Comment by antiochp** [comment]
> Could you remind me why we need this "minimal chance of key collision"?

If we see a reorg then we will reapply all outputs and the MMR positions will change.
Output `x` at pos `n` might actually be at pos `m` after the reorg.

This is going to cause problems if we use this index as part of `is_unspent` as we cannot assume the output at pos `n` is actually the output we expect it to be.

This also applies to the "header by height" index as we check coinbase maturity via this index.
If we see a reorg we need to make sure a coinbase output is actually as old as it appears to be (and the reorg can potentially change its age).

Basically these indexes are really useful but we need to be _absolutely_ sure they accurately reflect the "source of truth" data in the MMRs themselves.

The "pos to height" part of this index is going to be very susceptible to reorgs because we would be using "pos" as the key to the index and we will need to verify the pos is correct before relying on it.

Does this make sense?

**Comment by DavidBurkett** [comment]
@antiochp I don't understand why adding the additional 'header height' to the commitment index is a problem, since we already have an index from commitment to mmr index. That's already affected by reorgs, but we handle it fine. Header height wouldn't make this any worse.

**Comment by antiochp** [comment]
> I don't understand

If we use 2 associated indexes (A) and (B) as discussed above then its going to be tempting to jump straight to (B) and we need to be aware that this may not be reliable during a fork/reorg.
So the pros of using two indexes would be an easier migration for the existing index entries but at the cost of needing to maintain additional indexes.

This is not an issue if we migrate the existing index and add additional data to it.

**Comment by garyyu** [comment]
Directly index the `commitment` to `(output_pos, block_height)` is more straightforward, isn't it?
```
commitment -> (output_pos, block_height)
```
For database backward compatibility, we can migrate the database on node startup.

@antiochp What's the problem of this index solution?  or do we have some other backward compatibility problems here?

**Comment by antiochp** [comment]
Other than the migration step I think this will work fine.
The only other consideration here is we do not always _need_ the block_height when accessing this index, so its additional data that needs to be read and deserialized. But the commitment (the index key) is significantly larger than the two u64 values anyway so this is probably a non-issue.

**Comment by garyyu** [comment]
OK, Thanks @antiochp 
Let me try the one index firstly, and if we feel ugly later, we can switch to two index solution then.

Today, I tried to compare with the actual speed improvement.

---
Here is the version `1.1.0` performance on my laptop:

The test I did:
```
$ time curl -0 -XGET http://127.0.0.1:13413/v1/txhashset/outputs?start_index=200000\&max=1000 2>/dev/null > ~/tmp/output.1000.1.old

real	0m1.003s
user	0m0.009s
sys	0m0.011s

The log:
20190618 22:25:10.955 DEBUG grin_api::handlers::transactions_api - outputs: start_index: 200000, max: 1000
20190618 22:25:11.915 DEBUG grin_api::handlers::transactions_api - outputs: got 1000 outputs
```

Run it **2nd** time:
```
$ time curl -0 -XGET http://127.0.0.1:13413/v1/txhashset/outputs?start_index=200000\&max=1000 2>/dev/null > ~/tmp/output.1000.1.old

real	0m0.586s
user	0m0.010s
sys	0m0.012s

The log:
20190618 22:25:22.224 DEBUG grin_api::handlers::transactions_api - outputs: start_index: 200000, max: 1000
20190618 22:25:22.775 DEBUG grin_api::handlers::transactions_api - outputs: got 1000 outputs
```

Run it **3rd** time:
```
$ time curl -0 -XGET http://127.0.0.1:13413/v1/txhashset/outputs?start_index=200000\&max=1000 2>/dev/null > ~/tmp/output.1000.1.old

real	0m0.534s
user	0m0.010s
sys	0m0.013s

The log:

20190618 22:25:24.959 DEBUG grin_api::handlers::transactions_api - outputs: start_index: 200000, max: 1000
20190618 22:25:25.452 DEBUG grin_api::handlers::transactions_api - outputs: got 1000 outputs
```

The 2nd and 3rd time are much faster just because of the database cache. So we only need measure and compare the 1st call, or change with different `start_index`.

---

Here is the test result of the optimized version (with index `commitment -> (output_pos, block_height)`):

```
$ time curl -0 -XGET http://127.0.0.1:13413/v1/txhashset/outputs?start_index=200000\&max=1000 2>/dev/null > ~/tmp/output.1000.1

real    0m0.115s
user    0m0.012s
sys    0m0.014s
```

So, roughly 10 times speed gain!

**Comment by garyyu** [comment]
And I find the database caching is very helpful also, about 70% speed improvement. In case a Grin node working as a service node for a lot of wallet client, this will be helpful to improve the service performance.

Does anyone know how to tune the cache feature of lmdb?  with some configurations for example?

**Comment by antiochp** [comment]
The interesting thing about lmdb is there is no cache.
The entire db is basically a wrapper around mmap.

**Comment by garyyu** [comment]
> The interesting thing about lmdb is there is no cache.

👍
then we need figure out why the 1st call of `/v1/txhashset/outputs?start_index=200000\&max=1000` is always slower than next calls.

**Comment by antiochp** [comment]
There is no db cache. There _will_ be a page cache at the kernel layer.
The OS will page part of the file into memory, this will be slow the first time it is read from disk but then fast on subsequent reads from memory.
The point here is there is nothing to configure at the application or db level - its all delegated to the operating system kernel.

**Comment by garyyu** [comment]
Closed by #2903
